### ISSUE TITLE
cctail abruptly stops with no error messages

### ISSUE THREAD

**Issue by jaxoncrosmas** [newrelic-experimental/newrelic-logs-for-salesforce-commerce-cloud#2]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

I am using this solution to ingest many logs from Salesforce Commerce Cloud (~200k logs/hour). Ingesting this many logs seems to make the cctail application unstable. Every few hours, I will see on NewRelic that log ingestion has abruptly stopped. When I checked the container, everything appeared to be normal (the container was running and all applications inside the container were in a RUNNING state). Docker stats was also normal with no concerns regarding CPU, memory, or I/O.

The only thing that was not normal is the cctail application logs. I am running cctail in debug mode using the `CCTAIL_ARGS` environment variable (`CCTAIL_ARGS=-d`) as mentioned in the documentation. So the cctail application will log each time it polls a log file on Salesforce. I noticed that the cctail application was no longer logging its requests to Salesforce, despite cctail being in a RUNNING state. It also does not produce any error logs at the time of the crash, it just stops logging. The cctail application stopped logging at the same time that NewRelic stopped receiving Salesforce logs.

It seems like cctail is either crashing or being killed. I initially tested this container against my development environment which is generating far fewer logs, and I did not see this issue. This issue only showed itself once I started ingesting a larger number of logs.

## Environment Details

* Docker version 18.09.7, build 2d0083d
* Red Hat Enterprise Linux (version 7.9)

**Comment by jaxoncrosmas** [comment]
I have modified the cctail command in the `/etc/supervisord.conf` file to include the `--max-old-space-size` node option to increase the heap size of the cctail application.

```sh
command=sh -c "node --max-old-space-size=2048 cctail/dist/cctail.js %(ENV_CCTAIL_ARGS)s | awk '{ print strftime(\"cctail:\ %%Y-%%m-%%d\ %%H:%%M:%%S\ -\"), $0 }'"
```

I have not been observing the crashing behavior mentioned in the initial issue since I made this change. Will continue to monitor and update this issue with any findings.

**Comment by jaxoncrosmas** [comment]
I did some debugging to determine the root cause of this issue. I modified the `supervisor.conf` file to enable garbage collection trace on cctail (with node runtime option `--trace-gc`). Using this, I could see that the GC was filling up quite quickly and maxing out at around 2GB. This inevitably resulted in a `JavaScript Heap Out of Memory` error causing cctail to exit.

I modified the `supervisor.conf` file again to increase the heap size to 4GB (with node runtime option `--max-old-space-size`). This temporarily worked until the GC eventually maxed out at around 3.5GB which is when the system actually ran out of memory and killed the cctail process. I saw the following message in the syslog:

`Out of memory: Kill process 24993 (node) score 535 or sacrifice child`

I'm not sure if this is considered a memory leak or just a consequence of the design of cctail. Either way, cctail cannot currently handle this large number of logs. To solve this, I ended up creating multiple instances of this container and modifying the `log.conf.json` file in each container so that each container only ingests one or two log types. This way, the load is spread across multiple instances instead of processing this large number of logs on a single container.

Just in case anyone is using this solution to ingest a large number of logs and is facing this same issue, I recommend the solution above.
